flow charts: "-" in id's produce confusing error messages

Example json:
```json
{
  "nodes": [
    {
      "id": "one-one",
      "label": "one"
    },
    {
      "id": "two",
      "label": "two"
    }
  ],
  "edges": [
    ["one-one", "two"]
  ]
}
```
Resulting error message:
```java
java.lang.RuntimeException: Error: <stdin>: syntax error in line 23 near '-'

	at org.testingisdocumenting.znai.website.WebSite.throwParsingErrorMessage(WebSite.java:505)
	at org.testingisdocumenting.znai.website.WebSite.parseMarkupAndUpdateTocItemAndSearch(WebSite.java:497)
	at org.testingisdocumenting.znai.website.WebSite$$Lambda$53/00000000E2028020.accept(Unknown Source)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1082)
	at org.testingisdocumenting.znai.website.WebSite.parseMarkups(WebSite.java:458)
	at org.testingisdocumenting.znai.website.WebSite.parse(WebSite.java:192)
	at org.testingisdocumenting.znai.website.WebSite.parseAndDeploy(WebSite.java:182)
	at org.testingisdocumenting.znai.website.WebSite$Configuration.deployTo(WebSite.java:840)
	at org.testingisdocumenting.znai.cli.ZnaiCliApp.generateDocs(ZnaiCliApp.java:178)
	at org.testingisdocumenting.znai.cli.ZnaiCliApp.start(ZnaiCliApp.java:94)
	at org.testingisdocumenting.znai.cli.ZnaiCliApp.start(ZnaiCliApp.java:65)
	at org.testingisdocumenting.znai.cli.ZnaiCliApp.main(ZnaiCliApp.java:70)
Caused by: java.lang.RuntimeException: Error: <stdin>: syntax error in line 23 near '-'
	at org.testingisdocumenting.znai.diagrams.graphviz.InteractiveCmdGraphviz.readTill(InteractiveCmdGraphviz.java:59)
	at org.testingisdocumenting.znai.diagrams.graphviz.InteractiveCmdGraphviz.svgFromGv(InteractiveCmdGraphviz.java:46)
	at org.testingisdocumenting.znai.diagrams.graphviz.GraphvizEngine.diagramFromGv(GraphvizEngine.java:41)
	at org.testingisdocumenting.znai.diagrams.FlowChartIncludePlugin.process(FlowChartIncludePlugin.java:84)
	at org.testingisdocumenting.znai.parser.commonmark.MarkdownVisitor.handleIncludePlugin(MarkdownVisitor.java:215)
	at org.testingisdocumenting.znai.parser.commonmark.MarkdownVisitor.visit(MarkdownVisitor.java:147)
	at org.testingisdocumenting.znai.parser.commonmark.include.IncludeBlock.accept(IncludeBlock.java:43)
	at znaishaded.org.commonmark.node.AbstractVisitor.visitChildren(AbstractVisitor.java:137)
	at znaishaded.org.commonmark.node.AbstractVisitor.visit(AbstractVisitor.java:28)
	at znaishaded.org.commonmark.node.Document.accept(Document.java:7)
	at org.testingisdocumenting.znai.parser.commonmark.MarkdownParser.parsePartial(MarkdownParser.java:92)
	at org.testingisdocumenting.znai.parser.commonmark.MarkdownParser.parse(MarkdownParser.java:64)
	at org.testingisdocumenting.znai.website.WebSite.parseMarkupAndUpdateTocItemAndSearch(WebSite.java:485)
	... 11 more
```
There are definitely not 23 lines in this json file :)

--- a/znai-diagrams/src/main/java/org/testingisdocumenting/znai/diagrams/graphviz/GraphvizDiagram.java
+++ b/znai-diagrams/src/main/java/org/testingisdocumenting/znai/diagrams/graphviz/GraphvizDiagram.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2021 znai maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -21,10 +22,10 @@ import java.util.List;
 import java.util.Map;
 
 public class GraphvizDiagram {
-    private String id;
-    private String svg;
-    private Map<String, List<String>> stylesByNodeId;
-    private Map<String, Boolean> isInvertedTextColorByStyleId;
+    private final String id;
+    private final String svg;
+    private final Map<String, List<String>> stylesByNodeId;
+    private final Map<String, Boolean> isInvertedTextColorByStyleId;
 
     public GraphvizDiagram(String id, String svg, Map<String, List<String>> stylesByNodeId,
                            Map<String, Boolean> isInvertedTextColorByStyleId) {

--- a/znai-diagrams/src/main/java/org/testingisdocumenting/znai/diagrams/graphviz/gen/GraphvizFromJsonGen.java
+++ b/znai-diagrams/src/main/java/org/testingisdocumenting/znai/diagrams/graphviz/gen/GraphvizFromJsonGen.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2021 znai maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -15,6 +16,8 @@
  */
 
 package org.testingisdocumenting.znai.diagrams.graphviz.gen;
+
+import org.testingisdocumenting.znai.utils.StringUtils;
 
 import java.util.*;
 import java.util.function.Consumer;
@@ -157,7 +160,7 @@ public class GraphvizFromJsonGen {
     }
 
     private String generateNode(DiagramNode node) {
-        return node.getId() + " [label=\"" + generateNodeLabel(node) + "\"" +
+        return StringUtils.wrapInDoubleQuotes(node.getId()) + " [label=\"" + generateNodeLabel(node) + "\"" +
                 generateSizeAttributesIfRequired(node) + "];";
     }
 
@@ -201,7 +204,8 @@ public class GraphvizFromJsonGen {
     }
 
     private String generateEdge(DiagramEdge edge) {
-        return edge.getFromId() + " -> " + edge.getToId() +
+        return StringUtils.wrapInDoubleQuotes(edge.getFromId()) + " -> " +
+                StringUtils.wrapInDoubleQuotes(edge.getToId()) +
                 (!edge.getDirection().isEmpty() ? "[dir=" + edge.getDirection() + "];" : ";");
     }
 

--- a/znai-diagrams/src/main/java/org/testingisdocumenting/znai/diagrams/graphviz/meta/GraphvizDiagramWithMeta.java
+++ b/znai-diagrams/src/main/java/org/testingisdocumenting/znai/diagrams/graphviz/meta/GraphvizDiagramWithMeta.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2021 znai maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -17,6 +18,7 @@
 package org.testingisdocumenting.znai.diagrams.graphviz.meta;
 
 import org.testingisdocumenting.znai.utils.RegexpUtils;
+import org.testingisdocumenting.znai.utils.StringUtils;
 
 import java.util.*;
 import java.util.regex.Matcher;
@@ -33,7 +35,7 @@ public class GraphvizDiagramWithMeta {
     private final String preprocessed;
     private final String[] contentLines;
     private final Map<String, List<String>> stylesById;
-    private GraphvizShapeConfig metaConfig;
+    private final GraphvizShapeConfig metaConfig;
 
     public static GraphvizDiagramWithMeta create(GraphvizShapeConfig metaConfig, String diagramContent) {
         return new GraphvizDiagramWithMeta(metaConfig, diagramContent);
@@ -61,7 +63,7 @@ public class GraphvizDiagramWithMeta {
     }
 
     private String replaceAndExtractMeta(Matcher m) {
-        String id = m.group(1);
+        String id = StringUtils.removeQuotes(m.group(1));
         String label = m.group(2);
 
         List<String> styles = extractStyles(label);
@@ -72,7 +74,8 @@ public class GraphvizDiagramWithMeta {
             additionalProps = "," + additionalProps;
         }
 
-        return id + " [label=\"" + removeMeta(label) + "\"" + additionalProps + "];";
+        return StringUtils.wrapInDoubleQuotes(id) +
+                " [label=\"" + removeMeta(label) + "\"" + additionalProps + "];";
     }
 
     private String removeMeta(String label) {

--- a/znai-diagrams/src/test/groovy/org/testingisdocumenting/znai/diagrams/graphviz/GraphvizEngineTest.groovy
+++ b/znai-diagrams/src/test/groovy/org/testingisdocumenting/znai/diagrams/graphviz/GraphvizEngineTest.groovy
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2021 znai maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -29,12 +30,12 @@ class GraphvizEngineTest {
         def engine = new GraphvizEngine(shapeConfig).registerRuntime(dot)
 
          def diagram = engine.diagramFromGv("dot", "id", """digraph Simple {
-    main [label="mn [world]"];
+    "main-bridge" [label="mn [world]"];
     server [label="server [a]"];
 
-    server -> main;
+    server -> "main-bridge";
 }""")
 
-        assert diagram.stylesByNodeId == [main: ['world'], server: ['a']]
+        assert diagram.stylesByNodeId == ['main-bridge': ['world'], server: ['a']]
     }
 }

--- a/znai-diagrams/src/test/groovy/org/testingisdocumenting/znai/diagrams/graphviz/gen/GraphvizFromJsonGenTest.groovy
+++ b/znai-diagrams/src/test/groovy/org/testingisdocumenting/znai/diagrams/graphviz/gen/GraphvizFromJsonGenTest.groovy
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2021 znai maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -33,18 +34,18 @@ class GraphvizFromJsonGenTest {
                         ["n1", "n2"],
                         ["n1", "n2", "both"],
                 ]])
-
-        gv.should == "digraph Generated {\n" +
-                "rankdir=LR;\n" +
-                "bgcolor=\"#ffffff00\";\n" +
-                "node [shape=record; fontsize=10; margin=0.2; fontname=Helvetica];\n" +
-                "\n" +
-                "n1 [label=\"l1\"];\n" +
-                "n2 [label=\"l2\"];\n" +
-                "\n" +
-                "n1 -> n2;\n" +
-                "n1 -> n2[dir=both];\n" +
-                "}"
+        
+        gv.should == 'digraph Generated {\n' +
+                'rankdir=LR;\n' +
+                'bgcolor="#ffffff00";\n' +
+                'node [shape=record; fontsize=10; margin=0.2; fontname=Helvetica];\n' +
+                '\n' +
+                '"n1" [label="l1"];\n' +
+                '"n2" [label="l2"];\n' +
+                '\n' +
+                '"n1" -> "n2";\n' +
+                '"n1" -> "n2"[dir=both];\n' +
+                '}'
     }
 
     @Test
@@ -54,7 +55,7 @@ class GraphvizFromJsonGenTest {
                 edges: [["n", "n"]
                 ]])
 
-        gv.should contain("n [label=\"multi\\nline\"]")
+        gv.should contain('"n" [label="multi\\nline"]')
     }
 
     @Test
@@ -73,8 +74,8 @@ class GraphvizFromJsonGenTest {
                         ]]
                 ])
 
-        gv.should contain("n1 [label=\"l1\" fixedsize=true width=4 height=2]")
-        gv.should contain("n2 [label=\"l2\" fixedsize=true width=4 height=2]")
+        gv.should contain('"n1" [label=\"l1\" fixedsize=true width=4 height=2]')
+        gv.should contain('"n2" [label=\"l2\" fixedsize=true width=4 height=2]')
     }
 
     @Test
@@ -93,38 +94,38 @@ class GraphvizFromJsonGenTest {
                         ]]
                 ])
 
-        gv.should contain("n1 [label=\"l1\" fixedsize=true width=4 height=2]")
-        gv.should contain("n2 [label=\"l2\" fixedsize=true width=8 height=1]")
+        gv.should contain('"n1" [label="l1" fixedsize=true width=4 height=2]')
+        gv.should contain('"n2" [label="l2" fixedsize=true width=8 height=1]')
     }
 
     @Test
     void "applies meta information for highlighted items"() {
         generate([nodes: [[id: "n", label: "l", highlight: true]], edges: [["n", "n"]]])
-        gv.should contain("n [label=\"l[h]\"]")
+        gv.should contain('"n" [label="l[h]"]')
     }
 
     @Test
     void "applies meta information when color group is specified"() {
         generate([nodes: [[id: "n", label: "l", colorGroup: "b"]], edges: [["n", "n"]]])
-        gv.should contain("n [label=\"l[b]\"]")
+        gv.should contain('"n" [label="l[b]"]')
     }
 
     @Test
     void "applies meta information when shape is specified"() {
         generate([nodes: [[id: "n", label: "l", shape: "database"]], edges: [["n", "n"]]])
-        gv.should contain("n [label=\"l[database]\"]")
+        gv.should contain('"n" [label="l[database]"]')
     }
 
     @Test
     void "applies meta information when shape and color is specified"() {
         generate([nodes: [[id: "n", label: "l", colorGroup: "b", shape: "database"]], edges: [["n", "n"]]])
-        gv.should contain("n [label=\"l[b database]\"]")
+        gv.should contain('"n" [label="l[b database]"]')
     }
 
     @Test
     void "applies meta information when shape and highlight is specified"() {
         generate([nodes: [[id: "n", label: "l", highlight: true, shape: "database"]], edges: [["n", "n"]]])
-        gv.should contain("n [label=\"l[h database]\"]")
+        gv.should contain('"n" [label="l[h database]"]')
     }
 
     @Test
@@ -145,12 +146,12 @@ class GraphvizFromJsonGenTest {
                 'bgcolor="#ffffff00";\n' +
                 'node [shape=record; fontsize=10; margin=0.2; fontname=Helvetica];\n' +
                 '\n' +
-                'a [label="A"];\n' +
-                'b [label="B"];\n' +
-                'c [label="C"];\n' +
+                '"a" [label="A"];\n' +
+                '"b" [label="B"];\n' +
+                '"c" [label="C"];\n' +
                 '\n' +
-                'a -> b;\n' +
-                'c -> d;\n' +
+                '"a" -> "b";\n' +
+                '"c" -> "d";\n' +
                 '}'
     }
 

--- a/znai-diagrams/src/test/groovy/org/testingisdocumenting/znai/diagrams/graphviz/meta/GraphvizDiagramWithMetaTest.groovy
+++ b/znai-diagrams/src/test/groovy/org/testingisdocumenting/znai/diagrams/graphviz/meta/GraphvizDiagramWithMetaTest.groovy
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2021 znai maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -25,18 +26,18 @@ class GraphvizDiagramWithMetaTest {
     @Test
     void "should extract styles based on labels"() {
         def diagram = GraphvizDiagramWithMeta.create(shapeConfig, """digraph Simple {
-    main [label="mn [a b]"];
+    "main-entry" [label="mn [a b]"];
     server [label="server [a]"];
 
-    main -> server;
+    "main-entry" -> server;
 }
 """)
-        assert diagram.stylesById == [main: ['a', 'b'], server: ['a']]
+        assert diagram.stylesById == ["main-entry": ['a', 'b'], server: ['a']]
         Assert.assertEquals("digraph Simple {\n" +
-                "    main [label=\"mn\"];\n" +
-                "    server [label=\"server\"];\n" +
+                "    \"main-entry\" [label=\"mn\"];\n" +
+                "    \"server\" [label=\"server\"];\n" +
                 "\n" +
-                "    main -> server;\n" +
+                "    \"main-entry\" -> server;\n" +
                 "}", diagram.getPreprocessed())
     }
 
@@ -44,6 +45,6 @@ class GraphvizDiagramWithMetaTest {
     void "should add shape information for database style"() {
         def diagram = GraphvizDiagramWithMeta.create(shapeConfig, """main [label="mn [person a]"];""")
 
-        Assert.assertEquals("main [label=\"mn\",shape=octagon; width=1; height=2; fixedsize=true];", diagram.getPreprocessed())
+        Assert.assertEquals("\"main\" [label=\"mn\",shape=octagon; width=1; height=2; fixedsize=true];", diagram.getPreprocessed())
     }
 }

--- a/znai-docs/znai/release-notes/1.51/_fix-2021-07-03-flow-diagram-dash.md
+++ b/znai-docs/znai/release-notes/1.51/_fix-2021-07-03-flow-diagram-dash.md
@@ -1,0 +1,1 @@
+* Fix: handle dashes in [flow diagram](visuals/flow-diagrams) node ids 

--- a/znai-docs/znai/visuals/flow-diagrams/simple-dag-shapes.json
+++ b/znai-docs/znai/visuals/flow-diagrams/simple-dag-shapes.json
@@ -11,7 +11,7 @@
       "shape": "database"
     },
     {
-      "id": "anotherDb",
+      "id": "another-db",
       "label": "three input",
       "shape": "database-bottom-label"
     },
@@ -36,7 +36,7 @@
     ["researcher", "processor"],
     ["researcher", "config"],
     ["db", "processor"],
-    ["anotherDb", "processor"],
+    ["another-db", "processor"],
     ["config", "outcome"],
     ["processor", "outcome"]
   ]

--- a/znai-utils/src/main/java/org/testingisdocumenting/znai/utils/StringUtils.java
+++ b/znai-utils/src/main/java/org/testingisdocumenting/znai/utils/StringUtils.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2021 znai maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -80,6 +81,23 @@ public class StringUtils {
                 "");
     }
 
+    public static String removeQuotes(String text) {
+        if ((text.startsWith("\"") || text.startsWith("'")) &&
+                text.endsWith("\"") || text.endsWith("'")) {
+            return text.substring(1, text.length() - 1);
+        }
+
+        return text;
+    }
+
+    public static String wrapInDoubleQuotes(String text) {
+        if (text.startsWith("\"") && text.endsWith("\"")) {
+            return text;
+        }
+
+        return "\"" + text + "\"";
+    }
+
     private static String removeIndentation(String line, Integer indentation) {
         if (line.trim().isEmpty()) {
             return line;
@@ -96,7 +114,7 @@ public class StringUtils {
         int b = firstNonEmptyLineIdx(lines);
         int e = firstFromEndNonEmptyLineIdx(lines);
 
-        return lines.subList(b > 0 ? b : 0, e > 0 ? e + 1 : lines.size());
+        return lines.subList(Math.max(b, 0), e > 0 ? e + 1 : lines.size());
     }
 
     private static int firstNonEmptyLineIdx(List<String> lines) {
@@ -118,5 +136,4 @@ public class StringUtils {
 
         return -1;
     }
-
 }

--- a/znai-utils/src/test/groovy/org/testingisdocumenting/znai/utils/StringUtilsTest.groovy
+++ b/znai-utils/src/test/groovy/org/testingisdocumenting/znai/utils/StringUtilsTest.groovy
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2021 znai maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -58,5 +59,19 @@ line #_3\r""")
         Assert.assertEquals("a prefix:line1 line1\n" +
                 "         line2\n" +
                 "         line #3", concatenated)
+    }
+
+    @Test
+    void "remove quotes"() {
+        Assert.assertEquals("", StringUtils.removeQuotes('""'))
+        Assert.assertEquals("", StringUtils.removeQuotes("''"))
+        Assert.assertEquals("hello world", StringUtils.removeQuotes("'hello world'"))
+        Assert.assertEquals("hello world", StringUtils.removeQuotes('"hello world"'))
+    }
+
+    @Test
+    void "wrap in double quotes"() {
+        Assert.assertEquals('""', StringUtils.wrapInDoubleQuotes('""'))
+        Assert.assertEquals('"hello world"', StringUtils.wrapInDoubleQuotes("hello world"))
     }
 }

